### ISSUE TITLE
Issues with snake_casing in Python

### ISSUE THREAD

**Issue by lukehoban** [pulumi/pulumi#3151]
There are several intersecting issues with the current approach to automatic snake_casing for Python SDKs, which in total lead to:
1. Us not implementing what we're trying to do correctly
2. What we are trying to do being fundamentally "wrong"
3. Significant inconsistencies within and among providers

Our code for name translation in the Python SDK intends to implement the following: "For inputs, all properties, recursively, can be passed in either snake_case or camelCase.  For outputs, all properties, recursively, are snake_cased."

The act of "snake_casing" is done not based on an algorithm but based on a fixed table provided by each provider.  This table is "correctly" generated for Kubernetes, but is incorrectly (relative the the intention above) generated for TFGen.  In particular, for TFGen, only top level property names are added to the fixed table (along with keys of resources nested in `TypeMap`s which nearly never occur in Terraform - there are 5 cases total of this across AWS, Azure and GCP).

In practice, this means that for TFGen providers, the "actual" experience is effectively: "For inputs, top-level property names (projected as Python named constructor arguments) can be passed in either snake_case or camelCase, nested properties (projected as dicts) must be passed camelCase.  For outputs, top-level property names (projected as Python class properties) are available snake_cased,  and nested properties (projected as dicts) are available camelCased."  The Kubernetes provider implements the intention as described earlier.

The "intended design" is fundamentally broken though.  If implemented correctly, it causes user-defined dictionary keys passed to inputs like `Tags` to be transformed.  As it so happens, because we mostly don't implement it correctly for TFGen, we've avoided issues related to this.  (also because snake_case tag keys are likely relatively less common).

We believe that the "right" thing is actually this: "For inputs, top-level property names (projected as Python named constructor arguments) must be passed in snake_case, nested properties (projected as dicts) must be passed camelCase.  For outputs, top-level property names (projected as Python class properties) are available snake_cased,  and nested properties (projected as dicts) are available camelCased."  This matches a reasonable Python intuition that Python property and argument names are snake_case, but dict keys passed with quoted strings are not.

Properly "fixing" this would be a significant breaking change.  Even in our own code samples we appear to have almost every combination of things the above set of rules allows for.

But we believe we should take the following actions ASAP to get into a reasonable state where it is at least possible to recommend a consistent approach here (and apply it across all our examples and documentation), without a breaking change.

Specifically:
- [ ] Update Python output mapping to write outputs in both snake_case and in the form they come back naturally (camelCase) - warning and skipping if this leads to overwriting and existing property.
- [ ] Update API docs to always use camelCase names for nested inputs and outputs (these docs were just added this week)
- [ ] Update all examples in `pulumi/examples` and `pulumi/kubernetes` (and anywhere else) to use the style compatible with the "right" thing described above.
- [ ] Add a section on naming to https://www.pulumi.com/docs/intro/languages/python/ that describes how snake_casing vs. camelCasing works in general for all Python-based resource providers with some explicit examples (and possibly a note that the implementation is currently more flexible, but that this will likely be tightened up in the future).

Does that sound right @pgavlin @ellismg @lblackstone ?

**Comment by noliran** [comment]
I think I just ran into a side-effect of this behavior with `aws.cognito.IdentityProvider` (Python):

Pulumi translates `client_id`->`clientId` and `client_secret`->`clientSecret`, even though in this specific API (CreateIdentityProvider) AWS expects the children of `provider_details` to be snake cased in its API as shown [here](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_CreateIdentityProvider.html#CognitoUserPools-CreateIdentityProvider-request-ProviderDetails).

This effectively means that `aws.cognito.IdentityProvider` is broken in Pulumi when used with Python..

I'm using this workaround in the mean time but would love to know if there's any other way around this:
```
class PatchedIdentityProvider(aws.cognito.IdentityProvider):
    def translate_output_property(self, prop):
        if prop in ['client_id', 'client_secret']: return prop
        return aws._tables.CAMEL_TO_SNAKE_CASE_TABLE.get(prop) or prop

    def translate_input_property(self, prop):
        if prop in ['client_id', 'client_secret']: return prop
        return aws._tables.SNAKE_TO_CAMEL_CASE_TABLE.get(prop) or prop
```

**Comment by justinvp** [comment]
> I'm using this workaround in the mean time

Hi @noliran, sorry for the trouble. In the meantime, yes, this is a good way to work around the problem until we've addressed this more fundamentally.

Though, it would be better as:

```python
class PatchedIdentityProvider(aws.cognito.IdentityProvider):
    def translate_output_property(self, prop):
        if prop in ['client_id', 'client_secret']: return prop
        return super().translate_output_property(prop)

    def translate_input_property(self, prop):
        if prop in ['client_id', 'client_secret']: return prop
        return super().translate_input_property(prop)
```

Which avoids referring to the "private" `_tables` module, which could potentially change in the future.

**Comment by justinvp** [comment]
The snake_casing issues have been addressed in the 3.0-based provider Python SDKs with (https://github.com/pulumi/pulumi/pull/6695 and https://github.com/pulumi/pulumi/pull/6696).

Preview releases that include the fix are now available for the following providers:

- pulumi-kubernetes 3.0.0-beta.2
- pulumi-azure 4.0.0-beta.2
- pulumi-gcp 5.0.0-beta.2
- pulumi-azure-native 1.0.0-beta.2
- pulumi-random 4.0.0-beta.2
- pulumi-azuread 4.0.0-beta.2
- pulumi-aws 4.0.0-beta.2
